### PR TITLE
Revamp evolution tables

### PR DESCRIPTION
**Pull request by SadisticMystic**
The pokedex currently uses a lone "evoLevel" numeric field to validate all evolutions, which has its own limtations: while a Zubat would need to reach level 23 in order to evolve to Crobat, there are some games that allow you to catch Golbat directly at a level below 22, and in those cases it can evolve one level after it was caught, not necessarily having to wait for 23.

This adds "evoType" as a field, to begin a framework for validating those cases (but does not yet begin implementing the logic for them). "level" by itself represents the most common use case, where the evolution is triggered by leveling up to a certain minimum level. Additional constraints beyond the level threshold (Pangoro, Malamar, Solgaleo) are not expressed, nor should they ever need to be.

Beyond that, other values of evoType that begin with "level" (including levelHappy, levelMove, levelArea, etc.) represent conditions that have nothing to do with a minimum level threshold, but the act of evolving must necessarily consume a level-up anyway. By contrast, evoTypes of "trade" or "stone" can evolve without changing level.

Going along with that, I've catalogued the lowest available level for every species in every generation. Since hatched eggs are generally the most all-encompassing source for legality purposes, any encounter that doesn't beat the level you'd get with a hatched egg and/or natural evolution levels is not worth bothering to list. But I've added fields "lowestEncounter1", "lowestEncounter2", and so on to check for "underleveled" origins in generation 1, 2, and so on. More recent games have ability considerations linked to various sources, so the scheme gets more complex:

- lowestEncounter5 represents encounters that **cannot** have their hidden ability, or else the species doesn't have a distinct HA at all
- lowestEncounter5H represents encounter sources that will **always** have the hidden ability
- lowestEncounter6 and 7 (with no suffix) are **agnostic** about hidden ability or not, since more encounter types in those games are capable of spawning HA, and are thus "normal" for those generations
- 6N and 7N sources **cannot** have hidden abilities at the listed level
- 6E accounts for the peculiarity of DexNav: not only are these encounters agnostic about HA, but they can include **any one egg move** (but only one) despite not originating as an egg

At Zarel's request, I've stuck those minimum encounter levels in pokedex.js for the time being, but I'm not convinced that's where they're going to stay (it's not data intrinsic to the particular species, the way evolution methods are, just a by-product of the encounter tables in various games).

The last piece of data this adds is a "canHatch" field, which represents the potential for an egg to hatch directly as this species. Where this field is absent, it's implied to be true if the eggGroups contain neither Ditto nor Undiscovered, *and* the baseSpecies (or in its absence, the current species itself), has no "prevo" field; otherwise it can be assumed false. The purpose of the field is to account for cases like both Happiny (which needs the override to true despite being in Undiscovered, like all babies) and Chansey (which needs the override due to having a listed prevo, as the egg will hatch directly at Chansey stage instead of Happiny stage as long as you don't use incense). This may defy the precedent of Pomeg-glitch eggs (mostly in the context of old-gen LC) as far as what's considered a legal source, and might be worth a discussion before implementing any logic related to that field.

Competitively, this overhaul would have very minimal effect unless people are trying to make Pika Cup-like formats, but it's for completeness. Besides, people might like knowing that Ice Shard + Endeavor Donphan (which currently doesn't validate below level 25) can in fact be legally obtained by 24 thanks to the 6E source, and I'm sure some people have tried battling by plugging their in-game teams (which may include underleveled Pokemon that don't currently validate) into Showdown, rather than using in-game link battles for whatever reason. At least, once all the logic is fully implemented, it'll lead to fewer complaints from that group of people.

### CODE REVIEW

**Comment by Zarel** [comment]
I would not use an `evoType: "level"` field - that can just be automatically inferred from the existence of an `evoLevel` field.

For encounter levels... this might be a bit too much data for `pokedex.js` - maybe we want to stick it in `formats-data.js`?

File: dev-tools/globals.ts [hunk 1]
```diff
@@ -583,10 +584,22 @@ interface TemplateData {
 	evoLevel?: number
 	evoMove?: string
 	evos?: string[]
+	evoType?: string
```
Comment (Zarel) [inline dev-tools/globals.ts]: I would make this a string enum.

So:

`evoType?: "level" | "levelHappy" | "trade"` etc

**Comment by KrisXV** [comment]
would it be worthwhile to refactor evolutions into a new object, titled something like `evoData`?

**Comment by Zarel** [comment]
@KrisXV Hm, that might not be a bad idea.

`evolution: {from: "Bulbasaur", level: 16}`

`evolution: {from: "Magmar", item: "Magmarizer", trade: true},`

**Comment by KrisXV** [comment]
yeah pretty much what i was thinking

**Comment by viivgit** [comment]
I am not sure if you were done committing, but I noticed that some info that you entered in your first version of the changes (with `lowestEncounterX`), such as for the legendary birds (I just stopped after these ones, so there could be more), appear to have disappeared with the newer commits (with `encounters: [{...}]`).

It doesn't appear immediately obvious to me then why only the info for Gen I is left now, despite the fact that you had entered stuff for all gens for these mons (even though the info you had entered for Gen VI was wrong since the birds have been realeased in Gen VI with their HA, and that you had classified them as `lowestEncounter6N `instead of `lowestEncounter6`, but that's another problem).

**Comment by SadisticMystic** [comment]
It turns out the lowest encounters for 3, 4, etc for legendaries were already included in the event listings, whose scope apparently includes in-game events too. I don't know how VC validation is currently handled (other than the fact that, for instance, level 1 Inner Focus Raikou with Curse validates as legal, when that probably shouldn't be allowed); maybe after that logic is all in place, those encounters will be moved up with the events proper (as VC Mew already is, from what I see).

**Comment by Zarel** [comment]
Oh, you can't just subsume changes, you have to actually do the Git merge thing. I can probably do them for you though.

**Comment by Zarel** [comment]
@SadisticMystic afaict you're not done with this, but this seems useful so far, and I haven't seen progress in a while, so I merged in what you have so far.

Do you plan on working on this more?

**Comment by SadisticMystic** [comment]
I'm still not familiar enough with the codebase regarding team validation to be confident writing any of the logic to integrate these new fields into the check without breaking stuff, if anyone else wants to handle that and put this framework to some actual use.

**Comment by Zarel** [comment]
Of course. That's something I can do myself the next time I work on validation.

In the meantime I was asking about for instance splitting `levelHappiness` off from `levelExtra`, etc.

**Comment by Zarel** [comment]
(Also an `evoItem` field, with `evoType` reflecting how the item is used – stone, hold, etc)

**Pull request merged**